File: Editorial/139_Word_Break/CSharp/2_Top_Down_Dynamic_Programming.cs
public class Solution {
    private string s;
    private IList<string> wordDict;
    private int[] memo;

    private bool IsValid(int i) {
        if (i < 0) {
            return true;
        }

        if (memo[i] != -1) {
            return memo[i] == 1;
        }

        foreach (string word in wordDict) {
            // Handle out of bounds case
            if (i - word.Length + 1 < 0) {
                continue;
            }

            if (s.Substring(i - word.Length + 1, word.Length) == word &&
                IsValid(i - word.Length)) {
                memo[i] = 1;
                return true;
            }
        }

        memo[i] = 0;
        return false;
    }

    public bool WordBreak(string s, IList<string> wordDict) {
        this.s = s;
        this.wordDict = wordDict;
        this.memo = new int[s.Length];
        Array.Fill(this.memo, -1);
        return IsValid(s.Length - 1);
    }
}

File: Editorial/139_Word_Break/C/2_Top_Down_Dynamic_Programming.c
#define MAX_LENGTH 1000
int memo[MAX_LENGTH];
int memoLength;
char* s;
char** wordDict;
int wordDictLen;
int isValid(int i) {
    if (i < 0) {
        return 1;
    }
    if (memo[i] != -1) {
        return memo[i];
    }
    for (int j = 0; j < wordDictLen; j++) {
        int wordLen = strlen(wordDict[j]);
        if (i - wordLen + 1 < 0) {
            continue;
        }
        if (strncmp(wordDict[j], s + i - wordLen + 1, wordLen) == 0 &&
            isValid(i - wordLen)) {
            memo[i] = 1;
            return 1;
        }
    }
    memo[i] = 0;
    return 0;
}
int wordBreak(char* _s, char** _wordDict, int _wordDictLen) {
    memset(memo, -1, sizeof(memo));
    memoLength = strlen(_s);
    s = _s;
    wordDict = _wordDict;
    wordDictLen = _wordDictLen;
    return isValid(memoLength - 1);
}
